acpi/src/lib.rs

```diff
@@ -95,189 +100,173 @@ pub enum AcpiError {
+        if self.revision == 0 {
+            let num_tables = ((self.mapping.length as usize) - size_of::<SdtHeader>()) / size_of::<u32>();
+            // ### Safety: Table pointer is known-good for these offsets and types.
+            let tables_base = unsafe { self.mapping.virtual_start().as_ptr().add(1).cast::<u32>() };
 
-            let num_tables = (mapping.length as usize - mem::size_of::<SdtHeader>()) / mem::size_of::<u32>();
-            let tables_base =
-                ((mapping.virtual_start().as_ptr() as usize) + mem::size_of::<SdtHeader>()) as *const u32;
+            for offset in 0..num_tables {
+                // ### Safety: See above safety message.
+                let sdt_header_address = unsafe { tables_base.add(offset).read_unaligned() } as usize;
 
-            for i in 0..num_tables {
-                result.process_sdt(unsafe { *tables_base.add(i) as usize })?;
+                // ### Safety: `RSDT` guarantees its contained addresses to be valid.
+                let table_result = unsafe { read_table(self.handler.clone(), sdt_header_address) };
+                if table_result.is_ok() {
+                    return table_result;
+                }
             }
         } else {
-            /*
-             * ACPI Version 2.0+. It's a XSDT!
-             */
-            mapping.validate(sdt::Signature::XSDT)?;
-
-            let num_tables = (mapping.length as usize - mem::size_of::<SdtHeader>()) / mem::size_of::<u64>();
-            let tables_base =
-                ((mapping.virtual_start().as_ptr() as usize) + mem::size_of::<SdtHeader>()) as *const u64;
-
-            for i in 0..num_tables {
-                result.process_sdt(unsafe { *tables_base.add(i) as usize })?;
+            let num_tables = ((self.mapping.length as usize) - size_of::<SdtHeader>()) / size_of::<u64>();
+            // ### Safety: Table pointer is known-good for these offsets and types.
+            let tables_base = unsafe { self.mapping.virtual_start().as_ptr().add(1).cast::<u64>() };
+
+            for offset in 0..num_tables {
+                // ### Safety: See above safety message.
+                let sdt_header_address = unsafe { tables_base.add(offset).read_unaligned() } as usize;
+
+                // ### Safety: `RSDT` guarantees its contained addresses to be valid.
+                let table_result = unsafe { read_table(self.handler.clone(), sdt_header_address) };
+                if table_result.is_ok() {
+                    return table_result;
+                }
             }
         }
 
-        Ok(result)
+        Err(AcpiError::TableMissing(T::SIGNATURE))
     }
 
-    /// Construct an `AcpiTables` from a custom set of "discovered" tables. This is provided to allow the library
-    /// to be used from unconventional settings (e.g. in userspace), for example with a `AcpiHandler` that detects
-    /// accesses to specific physical addresses, and provides the correct data.
-    pub fn from_tables_direct(
```

Comment:
I might be wrong but have we just gotten rid of this? Iirc this was added for Redox so that they could construct a set of tables in userspace, but I don't think they ended up using this library (could be wrong). Not sure if we still want to supply functionality like this?

Reply:
Yeah, I'm not really sure how to allow such an implementation with the new `AcpiTables`. I'm not opposed to the functionality, but I can't imagine how to make it happen.